projects/components/src/download-json/download-json.component.ts

```diff
@@ -0,0 +1,59 @@
+import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
+import { IconType } from '@hypertrace/assets-library';
+import { IconSize } from '@hypertrace/components';
+import { Observable } from 'rxjs';
+import { ButtonSize, ButtonStyle } from '../button/button';
+
+@Component({
+  selector: 'ht-download-json',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  styleUrls: ['./download-json.component.scss'],
+  template: `
+    <div class="download-json">
+      <ht-button
+        *ngIf="!this.dataLoading"
+        class="download-json"
+        icon="${IconType.Download}"
+        display="${ButtonStyle.Text}"
+        size="${ButtonSize.Large}"
+        [htTooltip]="this.tooltip"
+        (click)="this.triggerDownload()"
+      ></ht-button>
+      <ht-icon *ngIf="this.dataLoading" icon="${IconType.Loading}" size="${IconSize.Large}"></ht-icon>
+    </div>
+  `
+})
+export class DownloadJsonComponent {
+  @Input()
+  public dataSource!: Observable<unknown>;
```

Comment:
I like that better strictly from an api perspective, but from a functional one the issue there is it's going to force a download every time we show an icon, which we don't want.

Reply:
With respect to exporting Trace as Json feature, we can either optimize on space by fetching only when user clicks on the button or we can optimize on time by fetching in advance utilizing the idle browser/network/server cycle to have the data in advance (in the background)

In the former case, the download prompt will probably start after 2-3 sec (or more) depending upon the query. In the latter case, we are only using an idle browser and network cycle and download is immediate. 

With my suggestion, there are two problems:
1. Hypothetically, if we have multiple download buttons then we are downloading too much data in advance
2. If the first fetch query fails, Users would not get a second try since the button wouldn't show up.

Reply:
>In the former case, the download prompt will probably start after 2-3 sec (or more) depending upon the query. In the latter case, we are only using an idle browser and network cycle and download is immediate.

- It's not an idle cycle, since it would be downloading on page load when we're trying to download the page's main data
- 99 times out of 100 (random numbers! but honestly, I think its actually >99) this is not going to be used. Exporting data is not a main use case, it's for debugging (or moving data into a different tool), and pre-downloading each time is just wasteful - both for the client mentioned above, but also for the server. And like you said, as we start adding downloads elsewhere in the product, we're forcing all of them to pre-download.
- That 2-3 seconds seems like a pretty big overestimate to me, but I guess depends on connection and the trace. Either way, having that button show up delayed after we do the download rather than immediately (but taking time to download once clicked) is just worse UX to me, and contrary to how downloads work in pretty much any other product.

Reply:
- By idle cycle I mean, this process will not affect or block the rendering of the main contents on the page. We do load widgets in a dashboard in a similar fashion. This will be another async call right?

- Only with this Trace JSON feature, we are downloading a data which is not already displayed in the UI. Contrary to this, all other usage for this button should always be using the data which is displayed in the associated table/list. So a separate network shouldn't be required.

- I agree that this may still be used very occasionally.

- Yes. the delay was some rough estimate which depends on connection/load/trace. 

-  I don't really agree with your points regarding the ux being bad for button showing up with delay. We do this currently at so many places. It is not going to cause any flicker or anything. By the time user look for it, it will be there. We can also show this with loader. (Please refer to Blocked events sheet for a similar ux)

That being said, I am fine with downloading on click time. The only concern I have is to avoid any unintended network call with any subscription inside this component.

Reply:
How do we address that?

Reply:
> By idle cycle I mean, this process will not affect or block the rendering of the main contents on the page. We do load widgets in a dashboard in a similar fashion. This will be another async call right?

Every network call we make, while yes async, will slow down others that are active, right? making 3 calls in parallel is faster than making 4. The tubes are only so big!

> Only with this Trace JSON feature, we are downloading a data which is not already displayed in the UI. Contrary to this, all other usage for this button should always be using the data which is displayed in the associated table/list. So a separate network shouldn't be required.

Future use cases presumably may have to fetch the data needed to download, too. Some cases it may be downloading the same data that's already displayed, but certainly case by case.

> I don't really agree with your points regarding the ux being bad for button showing up with delay. We do this currently at so many places. It is not going to cause any flicker or anything. By the time user look for it, it will be there. We can also show this with loader. (Please refer to Blocked events sheet for a similar ux)

We could put it there with a loader, but those cases are really places where we _need_ to load. For example, dropdowns, data panels etc. that are all reliant on data even if you don't interact with them.

> The only concern I have is to avoid any unintended network call with any subscription inside this component.

I'm not positive what you mean by this, but if you mean the fact that it would make a new network call on each click, that's not necessarily wrong and is again how almost every download button in the world works - if I download something once, then click the download button again, it will just try to download it again (another network call). That said, the component's parent can easily decide by sharing or not sharing a passed in observable whether it wants that behavior or not - the beauty of observables!

Reply:
> Every network call we make, while yes async, will slow down others that are active, right? making 3 calls in parallel is faster than making 4. The tubes are only so big!

There is not much UI rendering involved with this call. If we profile the page, i am confident the overall UI response time would only go down fractionally.

Reply:
> I'm not positive what you mean by this

refer to test `resubscribing to a debounced query after execution triggers a debounced new request`. If we are using this on security events table, I would share the same data$ with this component. Since there is a subscription from within this component, it would likely trigger a secondary network call. (Unless I remember to share the observable inside the parent component)

Reply:
> There is not much UI rendering involved with this call. If we profile the page, i am confident the overall UI response time would only go down fractionally.

It's not the UI rendering, it's the network bandwidth - you can only transmit so many bits, no matter how many calls you make.

> Example, If we are using this on events table, I would share the same data$ with this component. Since there is a subscription from within this component, it would likely trigger a secondary network call. (Unless I remember to share the observable inside the parent component)

If the parent wants to use the same observable in two places but only fetch once, its on the parent to share it - no one else can make that decision. Taking in a sync string for example and the parent using an async pipe like you suggested above `*ngIf="this.data$ | async as data"` would dupe the call too - except it would dupe it on every parent load, rather than just if the user clicked download. All of it is a bit moot considering these calls are going to be cached in apollo, but if anything it's another argument towards deferring the download IMO.